### ISSUE TITLE
Async setup work in API route

### ISSUE THREAD

**Issue by jgoux** [vercel/next.js#8632]
# Feature request

## Is your feature request related to a problem?

Currently it's not possible to make async operations before exporting the handler from a API route.

## Describe the solution you'd like

I'd like to be able to use the same mecanism from [micro](https://zeit.co/blog/micro-9-1) to be able to do async work before returning a handler.

## Additional context

I opened a discussion about it on spectrum : https://spectrum.chat/next-js/general/async-api-route~b1d7dddd-50fc-4f6f-a565-63fa68a725be

**Comment by huv1k** [comment]
Hey @jgoux, for questions we have [spectrum](https://spectrum.chat/next-js) and you already asked there. This is possible using a normal async function like this for example:
```js
export default async (req, res) => {
  await new Promise(resolve => {
    setTimeout(() => {
      resolve()
    }, 2000);
  })

  res.end('Test')
}
```

**Comment by timneutkens** [comment]
I was discussing this with @jgoux on twitter, it's more so related to https://zeit.co/blog/micro-9-1

You can already use async functions as outlined by @huv1k. It's just not possible to run a setup script on bootup

**Comment by jgoux** [comment]
Yes sorry I wasn't clear about it!

**Comment by timneutkens** [comment]
Added it to the roadmap

**Comment by Janpot** [comment]
Maybe, long term future, something could be done with a [top-level `await`](https://github.com/tc39/proposal-top-level-await)?

**Comment by andresgutgon** [comment]
Hi I want to do something similar. I want to initialize a GraphQL server with an async generation of my schema. 

I was playing a bit and this is kind of working:
``` typescript
import 'reflect-metadata'
import { GraphQLSchema } from 'graphql'
import {
  buildSchema,
  BuildSchemaOptions,
  Ctx,
  ObjectType,
  Field,
  Resolver,
  Query
} from 'type-graphql'
import { IncomingMessage, ServerResponse } from 'http'
import { ApolloServer } from 'apollo-server-micro'

@ObjectType('user')
class User { // This is just a JS class not interesting }

@Resolver()
class DummyResolver {
  @Query(() => User, { nullable: true })
  async me(@Ctx() _ctx: any): Promise<User> {
    const user: User = new User(1, 'bob@example.com')
    return Promise.resolve(user)
  }
}

const buildServer = async (): Promise<ApolloServer> => {
  const options: BuildSchemaOptions = { resolvers: [DummyResolver] }
  const schema: GraphQLSchema = await buildSchema(options)
  return new ApolloServer({ schema })
}

export const config = { api: { bodyParser: false } }
let server: ApolloServer | null = null

export default async (req: IncomingMessage, res: ServerResponse) => {
  const apolloServer: ApolloServer = server || await buildServer()
  return apolloServer.createHandler({
    path: '/api/graphql'
  })(req, res)
}
```

So the thing is that I look for `server` variable and if it's not present I call `buildServer` which return an `ApolloServer` with my schema. 

Do you think this is going in the right direction?

**Comment by fromi** [comment]
@andresgutgon Thanks for that, I was here to do exactly the same think.
As a minor improvment, I think you could call "createHandler" only once:
```
let apolloServerHandler: (req: any, res: any) => Promise<void>


const getApolloServerHandler = async () => {
  if (!apolloServerHandler) {
    const schema = await buildSchema({
      resolvers: [/* Your resolvers here */]
    })
    apolloServerHandler = new ApolloServer({schema}).createHandler({path: '/api/graphql'})
  }
  return apolloServerHandler
}

export default async (req: NextApiRequest, res: NextApiResponse) => {
  const apolloServerHandler = await getApolloServerHandler()
  return apolloServerHandler(req, res)
}

export const config = {api: {bodyParser: false}}
```

**Comment by andresgutgon** [comment]
I found another issue. When you request a NextJs page it goes this way:

```
-> /pages/my-next-page
        |
        |--- /my-next-page request from ApolloClient my data to /pages/api/graphql
```
This was generating an error because GraphQL schema was being generated multiple times and was getting this error:

```
Error: Schema must contain uniquely named types but contains multiple types named "users"
```

The solution I found was to cache in a `global` variable generated graphQL schema this way:

``` javascript
/**
 * We cache schema in a global variable
 * that lives in the server.
 * We use `globalThis`
 * Explained here:
 * https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/#type-checking-for-globalthis
 */
export default async function buildSchemaFn (): Promise<GraphQLSchema> {
  if (globalThis.schema) return globalThis.schema

  globalThis.schema = await buildSchema({
    resolvers: [DummyResolver]
  })

  return globalThis.schema
}
```

I hope this can help someone.

To be recap my problems are because I want to use:

1. New next `api routes`
2. `type-graphql` wich require an `asycn` graphql schema generation

If you are using another server (like Express) to build your graphql or if you're not using [type-graphql](https://github.com/MichalLytek/type-graphql) you should be fine following next example

**Comment by fromi** [comment]
I had the same issue, fixed by using the beta version of type-graphql which fixes the duplicate schema generation. ("type-graphql": "^0.18.0-beta.7" right now)

**Comment by xmflsct** [comment]
I have a similar issue. Transferred from Gatsby to Next.js, I am used to acquire CMS data before building pages.

In my Next.js project, I cache CMS data to a json file to prevent hitting the limit of API calls. But then when building pages, multiple instances happen at the same time while I am caching the json file.

**Comment by linosorice** [comment]
I have the same problem, I want to build a graphql schema at runtime on the server.

**Comment by good-idea** [comment]
This solution based on @andresgutgon's works for me, but I'm seeing that the server is re-built for every request. In my case, this setup is pretty quick so it's not a big deal, but, I'm wondering if there was some way to preserve the server between requests.

```
import { IncomingMessage, ServerResponse } from 'http'
import { ApolloServer } from 'apollo-server-micro'
import { getGraphqlServer } from './getGraphqlServer'

export const config = { api: { bodyParser: false } }
let server: ApolloServer | null = null

export default async (req: IncomingMessage, res: ServerResponse) => {
  console.log(Boolean(server))
  const apolloServer: ApolloServer = server || (await getGraphqlServer())
  server = apolloServer
  return apolloServer.createHandler({
    path: '/api/graphql',
  })(req, res)
}
```

In the console, I see `false` with every request that comes in, which means that the top-level `server` variable isn't being re-used between requests. This is only happening with `vercel dev`, though, I haven't tested it in production yet.

**Comment by nvme0** [comment]
@good-idea using @fromi's improvement I'm getting `false` initially and `true` thereafter until the `/api/graphql` page is rebuilt by the dev server.

**Comment by Mi3liX9** [comment]
How can I add type-orm with it?
